url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/render/Rasterizer2D.java

```diff
@@ -0,0 +1,458 @@
+                          (int)vb.x, (int)vb.y, color);
+        }
+    }
+
+    private static void fillPolygonProcessLine(Vertex2D va, Vertex2D vb,
+        double h,
+        ArrayListOfDoubles spanBuffer)
+    {
+        //-----------------------------------------------------------------
+        double dx;
+        double dy;
+        double dxdy;
+        double dydx;
+        double b;
+        double x;
+
+        dx = vb.x - va.x;
+        dy = vb.y - va.y;
+
+        if ( Math.abs(dx) > VSDK.EPSILON && 
+             Math.abs(dy/dx) <= 1 + VSDK.EPSILON ) {
+            // Slop between -1 and 1
+            dydx = dy/dx;
+            b = va.y - dydx*va.x;
+
+            if ( Math.abs(dydx) < VSDK.EPSILON ) {
+                // Horizontal line case
+                if ( Math.abs(va.y - h) <= 0.5 ) {
+                    // Horizontal line is at "h" height
+                    spanBuffer.add(va.x);
+                    spanBuffer.add(vb.x);
+                }
+            }
+            else {
+                // Non-horizontal line with slope between -1 and 1
+                x = (h - b) / dydx;
+                if ( (va.y <= h && vb.y >= h) ||
+                     (va.y >= h && vb.y <= h) ) {
+                    spanBuffer.add(x);
+                }
+            }
+        }
+
+        if ( Math.abs(dy) > VSDK.EPSILON && 
+             Math.abs(dx/dy) <= 1 ) {
+            // Slop between -1 and 1 with respect to y
+            dxdy = dx/dy;
+            b = va.x - dxdy*va.y;
+
+            if ( Math.abs(dxdy) < VSDK.EPSILON ) {
+                // Vertical line case
+                if ( (va.y <= h && vb.y >= h) ||
+                     (va.y >= h && vb.y <= h)  ) {
+                    // Horizontal line is at "h" height
+                    spanBuffer.add(va.x);
+                }
+            }
+            else {
+                // Non-vertical line with slope between -1 and 1
+                x = dxdy*h + b;
+                if ( (va.y <= h && vb.y >= h) ||
+                     (va.y >= h && vb.y <= h) ) {
+                    spanBuffer.add(x);
+                }
+            }
+        }
+    }
+
+    /**
+    Current polygon filling rasterizer (scan-line) algorithm is a NAIVE
+    implementation of the general macro-algorithm outlined at [FOLE1992].3.6.
+
+    This implementation is working but inefficient, due to the fact that
+    neither scanline coherence nor edge coherence is taken into account. This
+    means that for each scanline, all polygon edges are intersected 
+    analitically. This is the "brute-force technique" recommended to be
+    avoided in [FOLE1992].3.6.3, but it is provided as reference to
+    compare efficiency with other (clever) aproaches that makes use of
+    mid-point algoritms for intersection finding, or "active edge tables"
+    (AETs) for efficient edge-coherence based traversals.
+    */
+    public static void fillPolygon(Image img, Polygon2D p, RGBPixel color)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 22. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 232-284 into a separate method.